VDA-StorageNode-Registry/contracts/StorageNodeRegistry.sol

```diff
@@ -0,0 +1,545 @@
+
+
+    /**
+     * @notice datacenterId counter
+     * @dev starts from 1
+     */
+    CountersUpgradeable.Counter private _datacenterIdCounter;
+    /**
+     * @notice nodeId counter
+     * @dev starts from 1
+     */
+    CountersUpgradeable.Counter private _nodeIdCounter;
+
+    /**
+     * @notice Denominator for latitude & longitude values
+     */
+    uint8 public constant DECIMAL = 8;
+
+    /**
+     * @notice Additional information for a data center
+     * @dev Used internally inside the contract
+     * @param isActive True when added. False after removed
+     * @param nodeCount Number of connected storage nodes
+     */
+    struct DatacenterInfo {
+        bool isActive;
+        uint nodeCount;
+    }
+
+    error InvalidDatacenterName();
+    error InvalidCountryCode();
+    error InvalidRegionCode();
+    error InvalidLatitude();
+    error InvalidLongitude();
+    error InvalidDatacenterId(uint id);
+    error HasDependingNodes();
+    error InvalidEndpointUri();
+    error InvalidDIDAddress();
+    error InvalidUnregisterTime();
+
+    // constructor() {
+    //     _disableInitializers();
+    // }
+
+    /**
+     * @dev initializer of deployment
+     */
+    function initialize() initializer public {
+        __VDAVerificationContract_init();
+    }
+
+    /**
+     * @notice Check validity of country code
+     * @param countryCode Unique two-character string code
+     */
+    function validateCountryCode(string calldata countryCode) internal pure {
+        if (bytes(countryCode).length != 2) {
+            revert InvalidCountryCode();
+        }
+    }
+
+    /**
+     * @notice Check validity of region code
+     * @param regionCode Unique region string code
+     */
+    function validateRegionCode(string calldata regionCode) internal pure {
+        if (bytes(regionCode).length == 0) {
+            revert InvalidRegionCode();
+        }
+    }
+
+    /**
+     * @notice Check validity of latitude and longitude values
+     * @param lat Latitude
+     * @param long Longitude
+     */
+    function validateGeoPosition(int lat, int long) internal pure {
+        if ( lat < -90 * int(10 ** DECIMAL) || lat > 90 * int(10 ** DECIMAL)) {
+            revert InvalidLatitude();
+        }
+
+        if (long < -180 * int(10 ** DECIMAL) || long > 180 * int(10 ** DECIMAL)) {
+            revert InvalidLongitude();
+        }
+    }
+
+    /**
+     * @notice Check validity of datacenterId
+     * @dev `datacenterId` should be the one that was added by contract owner
+     * @param id datacenterId
+     */
+    function checkDatacenterIdExistance(uint id) internal view {
+        if (!_datacenterInfo[id].isActive) {
+            revert InvalidDatacenterId(id);
+        }
+    }
+
+    /**
+     * @dev see { IStorageNodeRegistry }
+     */
+    function addDatacenter(Datacenter calldata data) external payable onlyOwner override
+         returns(uint) {
+        {
+            if (bytes(data.name).length == 0) {
```

Comment:
Should we set a max length? 256 characters?

Do we force alpha-numeric with `-` for all `name` and `id` fields?

Reply:
No need to set max length.